toolchain/check/type_structure.cpp

```diff
@@ -281,6 +287,36 @@ class TypeStructureBuilder {
           }
           break;
         }
+        case CARBON_KIND(SemIR::TypeOfInst type_of): {
+          (void)type_of;
+          auto const_id = context_.constant_values().Get(inst_id);
+          // TODO: TypeOfInst should not be encountered in impl lookup with a
+          // template-dependent value, since impl lookup on such values should
+          // be deferred to type-checking the specific. So this should become a
+          // CARBON_FATAL code path. For now it is possible, though, to reach
+          // here and it results in a diagnostic.
+          //
+          // However, TypeOfInst can be determined to be a concrete value for a
+          // template value with a non-template dependent type. This currently
+          // does not happen though, the TypeOfInst is always template
+          // dependent. If it does in the future then we will want to use the
+          // concrete constant value instruction of `inst_id` in the type
+          // structure:
+          //   if (const_id.is_concrete()) {
+          //     PushInstId(context_.constant_values().GetInstId(const_id));
+          //   }
```

Comment:
I think we only see canonical constant type values here, and a `TypeOfInst` instruction will only be a canonical constant value if it's a template constant. (If its argument is an `InstValue`, `TypeOfInst` evaluates to that instruction's value; otherwise it itself defines a template constant.)

I think we probably want to treat a `TypeOfInst` as a `?` in the type structure, because it represents an unknown generic type, and should be a worse match than a known concrete constant type.

Reply:
Ok. I have made it so and removed the od TODO. I do think there's some room here for `template n:! i32` where I think you're saying the `TypeOfInst` would be `template` still because its argument is, but the *type* is not template-dependent here so we could look for that and avoid using `?` in that case?